Sources/Xcodeproj/pbxproj().swift

```diff
@@ -87,33 +87,91 @@ public func pbxproj(srcroot: AbsolutePath, projectRoot: AbsolutePath, xcodeprojP
-        
+
         // Contruct an array of (refId, path, bflId) tuples for all the source files in the model.  The reference id is for the PBXFileReference in the group hierarchy, and the build file id is for the PBXBuildFile in the CompileSources build phase.
         let sourceFileRefs = fileRefs(forModuleSources: module, srcroot: srcroot)
-        
-        // Make an array of all the source file reference ids to add to the main group.
-        let sourceRefIds = sourceFileRefs.map{ $0.refId }
 
-        // the “Project Navigator” group for this module
-        print("        \(module.groupReference) = {")
-        print("            isa = PBXGroup;")
-        print("            name = '\(module.name)';")
-        print("            path = '\(moduleRoot.relative(to: projectRoot).asString)';")
-        print("            sourceTree = '<group>';")
-        print("            children = (" + sourceRefIds.joined(separator: ", ") + ");")
-        print("        };")
+        // Hash of AbsolutePath of a group and reference id's of its children.
+        // Children can be either a source file or a nested group
+        var groupsChildren = [AbsolutePath: Set<String>]()
+
+        // reference id's of immediate children of this module group.
+        var topLevelRefs  = Set<String>()
 
         // the contents of the “Project Navigator” group for this module
         for fileRef in sourceFileRefs {
             let path = fileRef.path.relative(to: moduleRoot)
             print("        \(fileRef.refId) = {")
+            print("            name = \(fileRef.path.basename);")
             print("            isa = PBXFileReference;")
             print("            lastKnownFileType = \(module.fileType(forSource: path));")
-            print("            path = '\(fileRef.path.relative(to: moduleRoot).asString)';")
+            print("            path = '\(fileRef.path.relative(to: srcroot).asString)';")
+            print("            sourceTree = '<group>';")
+            print("        };")
+
+            // This source file is immediate children of module directory, ie., no nested folders
+            if path.dirname == "." {
+                topLevelRefs.insert(fileRef.refId)
+                continue
+            }
+
+
+            // Generate paths as follows:
+            // Example:
+            //    input: "MyModule/Foo/foo.swift"
+            //   output: ["MyModule",
+            //            "MyModule/Foo"]
+            //
+            let paths = [AbsolutePath](sequence(first: AbsolutePath(path.dirname, relativeTo: moduleRoot), next: { path in
+                let parent = AbsolutePath(path.dirname, relativeTo: moduleRoot)
+                return parent == moduleRoot ? nil : parent
+            }))
+
+            // If this source file is under a nested group, get the group ref id, else get the file ref id.
+            let groupRef = paths.filter{ $0.parentDirectory == moduleRoot }.first.map{ $0.groupReference(srcroot: srcroot) }
```

Comment:
Isn't this just `paths.last`, and shouldn't it never fail (given case above w.r.t. `path.dirname`)?